backup/metadata.go

```diff
@@ -137,38 +140,75 @@ func flatV0ToBackupMetadata(source flatMetadataV0, haNodes int) core.BackupMetad
 	}
 }
 
-const agentConfPattern = "var/lib/juju/agents/machine-*/agent.conf"
-
-func countHANodes(directory string) (int, error) {
-	matches, err := filepath.Glob(filepath.Join(directory, topLevelDir, agentConfPattern))
-	if err != nil {
-		return 0, errors.Trace(err)
-	}
-	if len(matches) != 1 {
-		return 0, errors.Errorf("expected one machine agent.conf, found %d: %#v", len(matches), matches)
+func eachBsonDoc(source io.Reader, callback func([]byte) error) error {
+	var size uint32
+	var buf bytes.Buffer
+	for {
+		// Each bson document starts with a 32-bit little-endian size.
+		err := binary.Read(source, binary.LittleEndian, &size)
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
+		buf.Grow(int(size))
+		err = binary.Write(&buf, binary.LittleEndian, size)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		_, err = io.CopyN(&buf, source, int64(size-4))
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Pass the bytes rather than unmarshalling so the callback
+		// can decide how (or whether) to unmarshal it.
+		err = callback(buf.Bytes())
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		buf.Reset()
 	}
-	agentConf, err := os.Open(matches[0])
+}
+
+const jobManageModel = 2
+
+func countHANodes(directory, modelUUID string) (int, error) {
```

Comment:
Can we prefer to count the controllerNodes collection? It will be more accurate going forwards as we want to remove jobs from machines.

However for older Jujus that don't have a controllerNodes collection, we will have to count the jobs.

Reply:
Ok - I'll have a look how that's represented in the db docs and do that.